flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/source/reader/LimitableDataIterator.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.source.reader;
+
+import javax.annotation.Nullable;
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.flink.source.DataIterator;
+import org.apache.iceberg.flink.source.FileScanTaskReader;
+import org.apache.iceberg.io.FileIO;
+
+class LimitableDataIterator<T> extends DataIterator<T> {
+  private final RecordLimiter limiter;
+
+  LimitableDataIterator(
+      FileScanTaskReader<T> fileScanTaskReader,
+      CombinedScanTask task,
+      FileIO io,
+      EncryptionManager encryption,
+      @Nullable RecordLimiter limiter) {
+    super(fileScanTaskReader, task, io, encryption);
+    this.limiter = limiter;
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (limiter != null && limiter.reachLimit()) {
+      return false;
+    }
+
+    return super.hasNext();
+  }
+
+  @Override
+  public T next() {
+    if (limiter != null) {
```

Comment:
maybe instead of having null checks everywhere we could have a NOOP limiter that just wouldn't do anything, wdyt? That way we wouldn't need null checks everywhere and you'd end up using either the Noop limiter or a normal one with a valid limit

Reply:
I have moved the check of the long limit value (non-positive for unlimited) inside the `RecordLimiter`. please see if this is inline with what you are thinking.